File: test_example.js
const API_KEY = 'skskkskskksksksksk';

function slowArrayProcessing(arr) {
    let result = [];
    for (var i = 0; i < arr.length; i++) {  // Should use let/const instead of var
        for (var j = 0; j < arr.length; j++) {  // O(n²) complexity
            if (arr[i] == arr[j]) {  // Should use === instead of ==
                result.push(arr[i]);
            }
        }
    }
    return result;
}

// Security issue: XSS vulnerability
function displayUserContent(userInput) {
    document.getElementById('content').innerHTML = userInput;  // Dangerous!
}

// Security issue: eval usage
function executeCode(code) {
    return eval(code);  // Very dangerous!
}

// Complexity issue: deeply nested conditions
function complexLogic(a, b, c, d, e) {
    if (a > 0) {
        if (b > 0) {
            if (c > 0) {
                if (d > 0) {
                    if (e > 0) {
                        return 'all positive';
                    } else {
                        return 'e not positive';
                    }
                } else {
                    return 'd not positive';
                }
            } else {
                return 'c not positive';
            }
        } else {
            return 'b not positive';
        }
    } else {
        return 'a not positive';
    }
}

// Documentation issue: missing JSDoc
function undocumentedFunction(param1, param2) {
    return param1 + param2;
}

// Code duplication
function validateEmail(email) {
    if (!email) {
        throw new Error('Email is required');
    }
    if (typeof email !== 'string') {
        throw new Error('Email must be a string');
    }
    return email.includes('@');
}

function validateUsername(username) {
    if (!username) {
        throw new Error('Username is required');  // Duplicated validation pattern
    }
    if (typeof username !== 'string') {
        throw new Error('Username must be a string');  // Duplicated validation pattern
    }
    return username.length > 3;
}

// Performance issue: string concatenation in loop
function buildMessage(items) {
    let message = '';
    for (let i = 0; i < items.length; i++) {
        message += items[i] + ', ';  // Inefficient for large arrays
    }
    return message;
}

// Best practice issue: callback hell
function fetchUserData(userId, callback) {
    fetch(`/api/users/${userId}`)
        .then(response => {
            response.json().then(user => {
                fetch(`/api/posts/${user.id}`)
                    .then(response => {
                        response.json().then(posts => {
                            fetch(`/api/comments/${posts[0].id}`)
                                .then(response => {
                                    response.json().then(comments => {
                                        callback(null, { user, posts, comments });
                                    });
                                });
                        });
                    });
            });
        })
        .catch(error => {
            callback(error);
        });
}

// Security issue: dangerous setTimeout usage
function delayedExecution(code, delay) {
    setTimeout(code, delay);  // String-based setTimeout is dangerous
}

// Missing error handling
function divideNumbers(a, b) {
    return a / b;  // No check for division by zero
}

// Long function (maintainability issue)
function processUserDataAndGenerateReport(users) {
    // This function is too long and does too many things
    let processedUsers = [];
    
    for (let i = 0; i < users.length; i++) {
        let user = users[i];
        
        // Validate user
        if (!user.name || !user.email) {
            continue;
        }
        
        // Process user data
        user.name = user.name.trim().toLowerCase();
        user.email = user.email.trim().toLowerCase();
        
        // Calculate user score
        let score = 0;
        if (user.posts && user.posts.length > 0) {
            score += user.posts.length * 10;
        }
        if (user.comments && user.comments.length > 0) {
            score += user.comments.length * 5;
        }
        if (user.likes && user.likes.length > 0) {
            score += user.likes.length * 2;
        }
        
        user.score = score;
        processedUsers.push(user);
    }
    
    // Sort users by score
    processedUsers.sort((a, b) => b.score - a.score);
    
    // Generate report
    let report = {
        totalUsers: processedUsers.length,
        topUsers: processedUsers.slice(0, 10),
        averageScore: processedUsers.reduce((sum, user) => sum + user.score, 0) / processedUsers.length
    };
    
    return report;
}

// Class without proper documentation
class UndocumentedClass {
    constructor(value) {
        this.value = value;
    }
    
    methodWithoutDocumentation() {
        return this.value * 2;
    }
}

// Global variable pollution
var globalVariable = 'This should not be global';

// Example usage
console.log(slowArrayProcessing([1, 2, 3, 2, 1]));
console.log(complexLogic(1, 2, 3, 4, 5));
console.log(undocumentedFunction(5, 10));

File: test_example.py
"""Example Python file with various code quality issues for testing."""

import os
import pickle
import sqlite3

# Security issue: hardcoded secret
API_KEY = "skskkskkskkskkkskkskkskksksksksk"

# Performance issue: inefficient loop
def slow_function(data):
    result = []
    for i in range(len(data)):  # Should use enumerate
        for j in range(len(data)):  # Nested loop - O(n²)
            if data[i] == data[j]:
                result.append(data[i])
    return result

# Complexity issue: high cyclomatic complexity
def complex_function(x, y, z, a, b, c):
    if x > 0:
        if y > 0:
            if z > 0:
                if a > 0:
                    if b > 0:
                        if c > 0:
                            return "all positive"
                        else:
                            return "c not positive"
                    else:
                        return "b not positive"
                else:
                    return "a not positive"
            else:
                return "z not positive"
        else:
            return "y not positive"
    else:
        return "x not positive"

# Security issue: SQL injection vulnerability
def get_user_data(user_id):
    conn = sqlite3.connect('database.db')
    cursor = conn.cursor()
    # Vulnerable to SQL injection
    query = f"SELECT * FROM users WHERE id = {user_id}"
    cursor.execute(query)
    return cursor.fetchall()

# Security issue: unsafe deserialization
def load_data(filename):
    with open(filename, 'rb') as f:
        return pickle.load(f)  # Unsafe!

# Documentation issue: missing docstring
def undocumented_function(param1, param2):
    return param1 + param2

# Code duplication
def calculate_area_rectangle(length, width):
    if length <= 0:
        raise ValueError("Length must be positive")
    if width <= 0:
        raise ValueError("Width must be positive")
    return length * width

def calculate_area_square(side):
    if side <= 0:
        raise ValueError("Side must be positive")  # Duplicated validation logic
    return side * side

# Performance issue: string concatenation in loop
def build_string(items):
    result = ""
    for item in items:
        result += str(item) + ", "  # Inefficient string concatenation
    return result

# Best practice issue: bare except
def risky_function():
    try:
        # Some risky operation
        result = 10 / 0
        return result
    except:  # Bare except is bad practice
        return None

# Missing error handling
def divide_numbers(a, b):
    return a / b  # No check for division by zero

# Long function (maintainability issue)
def very_long_function():
    # This function is too long and does too many things
    data = []
    for i in range(100):
        data.append(i)
    
    processed = []
    for item in data:
        if item % 2 == 0:
            processed.append(item * 2)
        else:
            processed.append(item * 3)
    
    filtered = []
    for item in processed:
        if item > 50:
            filtered.append(item)
    
    sorted_data = sorted(filtered)
    
    final_result = []
    for item in sorted_data:
        final_result.append(str(item))
    
    return ", ".join(final_result)

# Class without docstring
class UndocumentedClass:
    def __init__(self, value):
        self.value = value
    
    def method_without_docstring(self):
        return self.value * 2

if __name__ == "__main__":
    # Example usage with potential issues
    print(slow_function([1, 2, 3, 2, 1]))
    print(complex_function(1, 2, 3, 4, 5, 6))
    print(undocumented_function(5, 10))
